README.md

```diff
@@ -1,59 +1,321 @@
+
+> If you are using [TypeScript](https://typescriptlang.org) (recommended to use)
+> make sure to use `I18nContextFlavor` to extend your bot context in order for
+> typings to work correctly.
+
+### Without sessions
+
+When not using sessions, the locale is selected from the `language_code` of the
+user. Or if the custom [locale negotiator function](#locale-negotiation) returns
+a locale, then it will be used.
+
+```ts
+import { Bot, Context } from "https://deno.land/x/grammy/mod.ts";
+import { I18n, I18nContextFlavor } from "https://deno.land/x/.../mod.ts";
+
+// Create a new I18n instance.
+const i18n = new I18n({
+  defaultLocale: "en",
+});
+
+// Load locales from the `locales` directory.
+await i18n.loadLocalesDir("locales");
+
+// For proper typings and auto-completions in IDEs,
+// extend the `Context` using `I18nContextFlavor`.
+type MyContext = Context & I18nContextFlavor;
+
+// Create a bot as usual, but use the modified Context type.
+const bot = new Bot<MyContext>(""); // <- Put your bot token here
 
-```plaintext
-yaml and json are ok
-Example directory structure:
-├── locales
-│   ├── en.yaml
-│   ├── en-US.yaml
-│   ├── it.json
-│   └── ru.yaml
-└── bot.js
+// Tell the bot to use the middleware from the instance.
+// Remember to register this middleware before registering
+// other middlewares.
+bot.use(i18n.middleware());
+
+bot.command("start", async (ctx) => {
+  // Use the method `t` or `translate` from the context and pass
+  // in the message id (key) of the message you want to get.
+  await ctx.reply(ctx.t("greeting"));
+});
+
+// Start your bot
+bot.start();
 ```
 
-```js
-import {Bot, session} from 'grammy'
-import {I18n, pluralize} from '@grammyjs/i18n'
+### Using sessions
+
+If `useSession` is set to true, the plugin will try to read the
+`__language_code` from user's session data and use it, before falling back to
+language code received from the updates.
+
+> **NOTE**: You have to register the `session` middleware before registering
+> i18n's middleware. Otherwise, it won't be able to read or write from session.
+
+```ts
+import {
+  Bot,
+  Context,
+  session,
+  SessionFlavor,
+} from "https://deno.land/x/grammy/mod.ts";
+import { I18n, I18nContextFlavor } from "https://deno.land/x/.../mod.ts";
+
+interface SessionData {
+  __language_code: string;
+}
+
+type MyContext =
+  & Context
+  & SessionFlavor<SessionData>
+  & I18nContextFlavor;
 
 const i18n = new I18n({
-  defaultLanguageOnMissing: true, // implies allowMissing = true
-  directory: 'locales',
-  useSession: true,
-})
+  defaultLocale: "en",
+  useSession: true, // whether get/set in session
+});
+
+// Load locales from the `locales` directory.
+await i18n.loadLocalesDir("locales");
+
+const bot = new Bot<MyContext>(""); // <- Put your bot token here
+
+// Remember to register `session` middleware before registering
+// middleware of the i18n instance.
+bot.use(session({
+  initial: () => {
+    return { __language_code: "en" };
```

Comment:
Wouldn't it be better to not return any default lang code here?
```suggestion
    return { };
```
That way, the plugin can infer it automatically from `ctx.from.language_code` as a fallback to when no value is present on the session.

Reply:
@KnorpelSenf Done.

Reply:
@HeeroML What? Why? There is a `localeNegotiator` option in i18n config. Isn't that enough?

The locale is selected in the following order.

1. From custom `localeNegotiator`. (if any provided)
2. Reads from session (if it is enabled)
3. User's from.language_code (from TG updates)
4. Default locale.